tracer/test/Datadog.Trace.TestHelpers/CustomTestFramework.cs

```diff
@@ -256,28 +263,161 @@ protected override async Task<RunSummary> RunTestCaseAsync(IXunitTestCase testCa
+                }
 
+                DelayedMessageBus messageBus = null;
                 try
                 {
-                    var result = await base.RunTestCaseAsync(testCase);
+                    while (true)
+                    {
+                        attemptsRemaining--;
+                        messageBus = new DelayedMessageBus(MessageBus);
+
+                        // If this throws, we just let it bubble up, regardless of whether there's a retry, as this indicates an xunit infra issue
+                        var summary = await RunTest(messageBus);
+                        if (summary.Failed == 0 || attemptsRemaining == 0)
+                        {
+                            // No failures, or not allowed to retry
+                            return summary;
+                        }
+
+                        _diagnosticMessageSink.OnMessage(new DiagnosticMessage($"RETRYING: {test} ({attemptsRemaining} attempts remaining, {retryReason})"));
+                        var testFullName = $"{TestMethod.TestClass.Class.Name}.{testCase.DisplayName}";
+                        await SendMetric(_diagnosticMessageSink, "dd_trace_dotnet.ci.tests.retries", testFullName, retryReason);
+                    }
+                }
+                finally
+                {
+                    // need to dispose of the message bus to flush any messages
+                    messageBus?.Dispose();
+                }
+
+                async Task<RunSummary> RunTest(DelayedMessageBus messageBus)
+                {
+                    using var timer = new Timer(
+                        _ => _diagnosticMessageSink.OnMessage(new DiagnosticMessage($"WARNING: {test} has been running for more than 15 minutes")),
+                        null,
+                        TimeSpan.FromMinutes(15),
+                        Timeout.InfiniteTimeSpan);
+
+                    _diagnosticMessageSink.OnMessage(new DiagnosticMessage($"STARTED: {test}"));
+
+                    try
+                    {
+                        var result = await testCase.RunAsync(_diagnosticMessageSink, messageBus, _constructorArguments, new ExceptionAggregator(Aggregator), CancellationTokenSource);
+
+                        var status = result.Failed > 0 ? "FAILURE" : (result.Skipped > 0 ? "SKIPPED" : "SUCCESS");
+
+                        _diagnosticMessageSink.OnMessage(new DiagnosticMessage($"{status}: {test} ({result.Time}s)"));
+
+                        return result;
+                    }
+                    catch (Exception ex)
+                    {
+                        _diagnosticMessageSink.OnMessage(new DiagnosticMessage($"ERROR: {test} ({ex.Message})"));
+                        throw;
+                    }
+                }
+
+                async Task SendMetric(IMessageSink outputHelper, string metricName, string testFullName, string reason)
```

Comment:
This _mostly_ duplicates some code we already have in `ErrorHelpers` but it's different enough that I don't think it's worth trying to abstract it at this point. We can always revisit that decision later